### PR TITLE
Add ernie4.py for ERNIE-4.5 

### LINKED ISSUE #7668: [Feature] Ernie4.5 and Ernie4.5MoE Model Support
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

support Ernie4.5 and Ernie4.5MoE Model

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by solrex**
## Motivation

- To close: https://github.com/sgl-project/sglang/issues/7668
* Support [ERNIE-4.5 models](https://huggingface.co/collections/baidu/ernie-45-6861cd4c9be84540645f35c9)

## Modifications

* Add ernie4.py for ERNIE-4.5 (Ernie4_5_ForCausalLM and Ernie4_5_MoeForCausalLM)

## Benchmark

GSM8K (200) Benchmark (using sglang/benchmark/gsm8k/bench_sglang.py)

| Model                                          | GSM8K (200) |
|:-------------------------------|:------------|
| ERNIE-4.5-21B-A3B-PT            | 92.5%             |
| ERNIE-4.5-21B-A3B-PT (with MTP)  | 93.5%       |
| ERNIE-4.5-0.3B-PT                   | 10.5%               |

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by solrex** [comment]
GSM8K (200) Benchmark (using sglang/benchmark/gsm8k/bench_sglang.py)

| Model                                          | GSM8K (200) |
|:-------------------------------|:------------|
| ERNIE-4.5-21B-A3B-PT            | 92.5%             |
| ERNIE-4.5-21B-A3B-PT (with MTP)  | 93.5%       |
| ERNIE-4.5-0.3B-PT                   | 10.5%               |

**Comment by solrex** [comment]
MMLU Benchmark (using sglang/benchmark/mmlu/bench_sglang.py)

| Model                                          | MMLU |
|:-------------------------------|:------------|
| ERNIE-4.5-21B-A3B-PT            | 74.1%             |
| ERNIE-4.5-21B-A3B-PT (with MTP)  | 73.8%       |
| ERNIE-4.5-0.3B-PT                   | 0%               |

```bash
# ERNIE-4.5-21B-A3B-PT
$ python3 bench_sglang.py --nsub 10
subject: abstract_algebra, #q:100, acc: 0.540
subject: anatomy, #q:135, acc: 0.741
subject: astronomy, #q:152, acc: 0.875
subject: business_ethics, #q:100, acc: 0.730
subject: clinical_knowledge, #q:265, acc: 0.842
subject: college_biology, #q:144, acc: 0.931
subject: college_chemistry, #q:100, acc: 0.410
subject: college_computer_science, #q:100, acc: 0.710
subject: college_mathematics, #q:100, acc: 0.540
subject: college_medicine, #q:173, acc: 0.757
Total latency: 47.252
Average accuracy: 0.741

# ERNIE-4.5-21B-A3B-PT (with MTP)
$ python3 bench_sglang.py --nsub 10
subject: abstract_algebra, #q:100, acc: 0.520
subject: anatomy, #q:135, acc: 0.726
subject: astronomy, #q:152, acc: 0.868
subject: business_ethics, #q:100, acc: 0.730
subject: clinical_knowledge, #q:265, acc: 0.838
subject: college_biology, #q:144, acc: 0.931
subject: college_chemistry, #q:100, acc: 0.420
subject: college_computer_science, #q:100, acc: 0.710
subject: college_mathematics, #q:100, acc: 0.530
subject: college_medicine, #q:173, acc: 0.775
Total latency: 40.049
Average accuracy: 0.738

# ERNIE-4.5-0.3B-PT 
$ python3 bench_sglang.py --nsub 10
subject: abstract_algebra, #q:100, acc: 0.000
subject: anatomy, #q:135, acc: 0.000
subject: astronomy, #q:152, acc: 0.000
subject: business_ethics, #q:100, acc: 0.000
subject: clinical_knowledge, #q:265, acc: 0.000
subject: college_biology, #q:144, acc: 0.000
subject: college_chemistry, #q:100, acc: 0.000
subject: college_computer_science, #q:100, acc: 0.000
subject: college_mathematics, #q:100, acc: 0.000
subject: college_medicine, #q:173, acc: 0.000
Total latency: 16.330
Average accuracy: 0.000
```

File: python/sglang/srt/models/ernie4.py [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+            if "mtp" in name:
+                continue
+            if "e_score_correction_bias" in name:
+                continue
+            for param_name, weight_name, shard_id in self.stacked_params_mapping:
+                if weight_name not in name:
+                    continue
+                if ("mlp.experts." in name) and name not in params_dict:
+                    continue
+                name = name.replace(weight_name, param_name)
```
[the remark this answers is not in the document]
Comment (solrex) [reply python/sglang/srt/models/ernie4.py:389]: The reasoning for this is provided in the comments.

**Comment by solrex** [comment]
After above changes, MTP acceptance rate improved to **50-90%**. Accuracy of A3B also saw a minor uplift, which aligns with [ERNIE's claimed accuracy](https://github.com/PaddlePaddle/ERNIE?tab=readme-ov-file#performance-of-post-trained-model-ernie-45-21b-a3b).

| Model                                          | MMLU |
|:-------------------------------|:------------|
| ERNIE-4.5-21B-A3B-PT            | 76.7%             |
| ERNIE-4.5-21B-A3B-PT (with MTP)  | 76.8%       |

**Comment by solrex** [comment]
As of today (July 3, 2025), the weight shape for `model.mtp_linear_proj.0.weight` in the **baidu/ERNIE-4.5-300B-A47B-PT** model is incorrect. Therefore, when loading **A47B** MTP layer, you should  transpose it with patch code bellow, otherwise, MTP loading will fail. (This issue did not occur on the **A3B** model.) 

Hope Baidu will fix this weight bug in future updates.

```python
diff --git a/python/sglang/srt/models/ernie4_mtp.py b/python/sglang/srt/models/ernie4_mtp.py
index 7fdbb51e..a2cbc121 100644
--- a/python/sglang/srt/models/ernie4_mtp.py
+++ b/python/sglang/srt/models/ernie4_mtp.py
@@ -175,6 +175,8 @@ class Ernie4_5_MoeForCausalLMMTP(nn.Module):
                     weight_loader = getattr(
                         param, "weight_loader", default_weight_loader
                     )
+                    if name.startswith("model.mtp_linear_proj"):
+                        loaded_weight = loaded_weight.transpose(0, 1)
                     weight_loader(param, loaded_weight)
                 else:
                     raise KeyError(f"Parameter '{name}' not found in MTP model.")
```

**Comment by solrex** [comment]
Hi @Hanrui-Wang, would you mind taking a look at this when you have some time? Thanks in advance!

**Comment by FlamingoPg** [comment]
> Hi @Hanrui-Wang, would you mind taking a look at this when you have some time? Thanks in advance!

Hi, I will review this PR.

**Comment by FlamingoPg** [comment]
> MMLU Benchmark (using sglang/benchmark/mmlu/bench_sglang.py)
> 
> Model	MMLU
> ERNIE-4.5-21B-A3B-PT	74.1%
> ERNIE-4.5-21B-A3B-PT (with MTP)	73.8%
> ERNIE-4.5-0.3B-PT	0%
> ```shell
> # ERNIE-4.5-21B-A3B-PT
> $ python3 bench_sglang.py --nsub 10
> subject: abstract_algebra, #q:100, acc: 0.540
> subject: anatomy, #q:135, acc: 0.741
> subject: astronomy, #q:152, acc: 0.875
> subject: business_ethics, #q:100, acc: 0.730
> subject: clinical_knowledge, #q:265, acc: 0.842
> subject: college_biology, #q:144, acc: 0.931
> subject: college_chemistry, #q:100, acc: 0.410
> subject: college_computer_science, #q:100, acc: 0.710
> subject: college_mathematics, #q:100, acc: 0.540
> subject: college_medicine, #q:173, acc: 0.757
> Total latency: 47.252
> Average accuracy: 0.741
> 
> # ERNIE-4.5-21B-A3B-PT (with MTP)
> $ python3 bench_sglang.py --nsub 10
> subject: abstract_algebra, #q:100, acc: 0.520
> subject: anatomy, #q:135, acc: 0.726
> subject: astronomy, #q:152, acc: 0.868
> subject: business_ethics, #q:100, acc: 0.730
> subject: clinical_knowledge, #q:265, acc: 0.838
> subject: college_biology, #q:144, acc: 0.931
> subject: college_chemistry, #q:100, acc: 0.420
> subject: college_computer_science, #q:100, acc: 0.710
> subject: college_mathematics, #q:100, acc: 0.530
> subject: college_medicine, #q:173, acc: 0.775
> Total latency: 40.049
> Average accuracy: 0.738
> 
> # ERNIE-4.5-0.3B-PT 
> $ python3 bench_sglang.py --nsub 10
> subject: abstract_algebra, #q:100, acc: 0.000
> subject: anatomy, #q:135, acc: 0.000
> subject: astronomy, #q:152, acc: 0.000
> subject: business_ethics, #q:100, acc: 0.000
> subject: clinical_knowledge, #q:265, acc: 0.000
> subject: college_biology, #q:144, acc: 0.000
> subject: college_chemistry, #q:100, acc: 0.000
> subject: college_computer_science, #q:100, acc: 0.000
> subject: college_mathematics, #q:100, acc: 0.000
> subject: college_medicine, #q:173, acc: 0.000
> Total latency: 16.330
> Average accuracy: 0.000
> ```

Are there any problem for 0.3B model 🤔, It looks accuracy is 0

**Comment by solrex** [comment]
> > MMLU Benchmark (using sglang/benchmark/mmlu/bench_sglang.py)
> > Model	MMLU
> > ERNIE-4.5-21B-A3B-PT	74.1%
> > ERNIE-4.5-21B-A3B-PT (with MTP)	73.8%
> > ERNIE-4.5-0.3B-PT	0%
> > ```shell
> > # ERNIE-4.5-21B-A3B-PT
> > $ python3 bench_sglang.py --nsub 10
> > subject: abstract_algebra, #q:100, acc: 0.540
> > subject: anatomy, #q:135, acc: 0.741
> > subject: astronomy, #q:152, acc: 0.875
> > subject: business_ethics, #q:100, acc: 0.730
> > subject: clinical_knowledge, #q:265, acc: 0.842
> > subject: college_biology, #q:144, acc: 0.931
> > subject: college_chemistry, #q:100, acc: 0.410
> > subject: college_computer_science, #q:100, acc: 0.710
> > subject: college_mathematics, #q:100, acc: 0.540
> > subject: college_medicine, #q:173, acc: 0.757
> > Total latency: 47.252
> > Average accuracy: 0.741
> > 
> > # ERNIE-4.5-21B-A3B-PT (with MTP)
> > $ python3 bench_sglang.py --nsub 10
> > subject: abstract_algebra, #q:100, acc: 0.520
> > subject: anatomy, #q:135, acc: 0.726
> > subject: astronomy, #q:152, acc: 0.868
> > subject: business_ethics, #q:100, acc: 0.730
> > subject: clinical_knowledge, #q:265, acc: 0.838
> > subject: college_biology, #q:144, acc: 0.931
> > subject: college_chemistry, #q:100, acc: 0.420
> > subject: college_computer_science, #q:100, acc: 0.710
> > subject: college_mathematics, #q:100, acc: 0.530
> > subject: college_medicine, #q:173, acc: 0.775
> > Total latency: 40.049
> > Average accuracy: 0.738
> > 
> > # ERNIE-4.5-0.3B-PT 
> > $ python3 bench_sglang.py --nsub 10
> > subject: abstract_algebra, #q:100, acc: 0.000
> > subject: anatomy, #q:135, acc: 0.000
> > subject: astronomy, #q:152, acc: 0.000
> > subject: business_ethics, #q:100, acc: 0.000
> > subject: clinical_knowledge, #q:265, acc: 0.000
> > subject: college_biology, #q:144, acc: 0.000
> > subject: college_chemistry, #q:100, acc: 0.000
> > subject: college_computer_science, #q:100, acc: 0.000
> > subject: college_mathematics, #q:100, acc: 0.000
> > subject: college_medicine, #q:173, acc: 0.000
> > Total latency: 16.330
> > Average accuracy: 0.000
> > ```
> 
> Are there any problem for 0.3B model 🤔, It looks accuracy is 0

@yinfan98 No. I tested it with FastDeploy and VLLM. They all report 0 with the 0.3B model. The MMLU benchmark requires the response to be 1 token among A, B, C, or D. If the test script's instructions are not followed properly—for example, if the answer is not provided in the first token—the results will be very poor. Therefore, the results are highly dependent on the instructions in the test script.

**Comment by solrex** [comment]
@yinfan98 Could you please let me know if there’s anything else you’d like me to address?

**Comment by finetunej** [comment]
Great to see this PR. Is support for the ERNIE 4.5 VL models also planned? The model performs very well, so it would be useful.

**Comment by solrex** [comment]
> Great to see this PR. Is support for the ERNIE 4.5 VL models also planned? The model performs very well, so it would be useful.

Sorry, this PR does not plan to support the ERNIE 4.5 VL model.

**Comment by solrex** [comment]
@zhyncs  Hi, just wondering if it would be possible to reassign a reviewer, as there hasn't been any feedback for a while. Thanks!

File: python/sglang/srt/models/ernie4.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+            self.shared_experts = Ernie4MLP(
+                hidden_size=config.hidden_size,
+                intermediate_size=intermediate_size,
+                hidden_act=config.hidden_act,
+                quant_config=quant_config,
+                reduce_results=False,
+                prefix=add_prefix("shared_experts", prefix),
+                **(
+                    {"tp_rank": 0, "tp_size": 1}
+                    if global_server_args_dict["enable_deepep_moe"]
```
Comment (ch-wan) [inline python/sglang/srt/models/ernie4.py]: this arg is deprecated

File: python/sglang/srt/models/ernie4_mtp.py
Comment (ch-wan) [inline python/sglang/srt/models/ernie4_mtp.py:1]: how about rename this file as *_eagle.py? This is what we did for deepseek/llama/qwen.

**Review by ch-wan — COMMENTED** [review]
Thank you for bringing ernie4 to SGLang. I left some quick comments.

File: python/sglang/srt/models/ernie4.py (hunk 2 above)
Comment (solrex) [reply python/sglang/srt/models/ernie4.py]: Removed.

File: python/sglang/srt/models/ernie4_mtp.py
Comment (solrex) [reply python/sglang/srt/models/ernie4_mtp.py:1]: Renamed.

File: python/sglang/srt/models/ernie4.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        super().__init__()
+        self.layer_id = layer_id
+        self.tp_size = get_tensor_model_parallel_world_size()
+        self.moe_num_shared_experts = getattr(config, "moe_num_shared_experts", 0)
+
+        if self.tp_size > config.moe_num_experts:
+            raise ValueError(
+                f"Tensor parallel size {self.tp_size} is greater than "
+                f"the number of experts {config.moe_num_experts}."
+            )
```
Comment (ch-wan) [inline python/sglang/srt/models/ernie4.py]: Do we need this check?

Comment (solrex) [reply python/sglang/srt/models/ernie4.py]: Thank you. It can be removed.
